Add pre upgrade/downgrade orchestrate

TeddyAndrieux opened the pull request:
**Component**:

'salt', 'upgrade', 'downgrade'
<!-- E.g. 'salt', 'containers', 'kubernetes', 'build', 'tests'... -->

**Context**: 

In #2110 we downgrade bootstrap before calling the orchestrate but we still need a state to cleanup some stuff

**Summary**:

Add a `pre` state for upgrade and downgrade to do all needed cleanup before upgrading or downgrading

**Acceptance criteria**: 

Working upgrade and downgrade

---

TeddyAndrieux committed:
salt: Make the downgrade orchestrate backward compatible

The downgrade orchestrate of the current version is called with a
salt-master already in the `dest_version` so the orchestrate need to be
backward compatible

TeddyAndrieux committed:
salt,build,script: Add pre downgrade orchestrate

Add dedicated orchestrate used to do some cleanup before downgrading the
metalk8s cluster. This orchestrate is called at the very beginning of
the downgrade, just before downgrade of the salt-master

TeddyAndrieux committed:
salt,build,script: Add pre upgrade orchestrate

Add dedicated orchestrate used to do some cleanup before upgrading the
metalk8s cluster. This orchestrate is called at the very beginning of
the upgrade, just after upgrade of the salt-master

bert-e commented:
# Hello teddyandrieux,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/2.5/bugfix/add-pre-upgrade-downgrade-orchestrate` with contents from `bugfix/add-pre-upgrade-downgrade-orchestrate`
and `development/2.5`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/2.5/bugfix/add-pre-upgrade-downgrade-orchestrate origin/development/2.5
 $ git merge origin/bugfix/add-pre-upgrade-downgrade-orchestrate
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.5/bugfix/add-pre-upgrade-downgrade-orchestrate
```

TeddyAndrieux commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/2.5/bugfix/add-pre-upgrade-downgrade-orchestrate` with contents from `bugfix/add-pre-upgrade-downgrade-orchestrate`
and `development/2.5`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/2.5/bugfix/add-pre-upgrade-downgrade-orchestrate origin/development/2.5
 $ git merge origin/bugfix/add-pre-upgrade-downgrade-orchestrate
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.5/bugfix/add-pre-upgrade-downgrade-orchestrate
```

TeddyAndrieux commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/2.5/bugfix/add-pre-upgrade-downgrade-orchestrate` with contents from `bugfix/add-pre-upgrade-downgrade-orchestrate`
and `development/2.5`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/2.5/bugfix/add-pre-upgrade-downgrade-orchestrate origin/development/2.5
 $ git merge origin/bugfix/add-pre-upgrade-downgrade-orchestrate
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.5/bugfix/add-pre-upgrade-downgrade-orchestrate
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort

NicolasT commented on scripts/upgrade.sh.in:
```diff
@@ -149,4 +160,5 @@ run "Performing Pre-Upgrade checks" precheck_upgrade
 [ $DRY_RUN -eq 1 ] && exit 0
 run "Upgrading bootstrap" upgrade_bootstrap
 run "Setting cluster version to $DESTINATION_VERSION" patch_kubesystem_namespace
+run "Launching the pre-upgrade" launch_pre_upgrade
```
Can't we do all this stuff in some orchestration rather than Bash?

TeddyAndrieux replied:
For upgrade we could but for downgrade we can't, because for downgrade we want to launch this "pre-downgrade" states before downgrading the bootstrap, 
Just prefer to have the same approach for upgrade and downgrade

TeddyAndrieux committed:
salt: apiserver cert localhost only kill apiserver onchanges

In salt states to add `localhost` to apiserver server certs only kill
apiserver to restart it if the cert changed

TeddyAndrieux committed:
salt: Use `version` label from object in pre-upgrade

Since #2132 we have a version label in each and every object we deploy
using metalk8s kubernetes module, use this version to know if we need to
remove nginx-ingress DaemonSet and Deployment before upgrading

bert-e commented:
# History mismatch



Merge commit #21d9b778d562988b8cecd4afaa172645c550c763 on the integration branch
`w/2.5/bugfix/add-pre-upgrade-downgrade-orchestrate` is merging a branch which is neither the current
branch `bugfix/add-pre-upgrade-downgrade-orchestrate` nor the development branch
`development/2.5`.

It is likely due to a rebase of the branch `bugfix/add-pre-upgrade-downgrade-orchestrate` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

TeddyAndrieux commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/2.5/bugfix/add-pre-upgrade-downgrade-orchestrate` with contents from `bugfix/add-pre-upgrade-downgrade-orchestrate`
and `development/2.5`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/2.5/bugfix/add-pre-upgrade-downgrade-orchestrate origin/development/2.5
 $ git merge origin/bugfix/add-pre-upgrade-downgrade-orchestrate
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.5/bugfix/add-pre-upgrade-downgrade-orchestrate
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort

Ebaneck commented on salt/metalk8s/kubernetes/apiserver/pre-upgrade.sls:
```diff
@@ -0,0 +1,11 @@
+{%- set cp_nodes = salt.metalk8s.minions_by_role('master') | sort %}
+
+# Always run this state as it do not change anything if cert are already good
```
Could read:
`Always run this state as it does not change anything if certs are already good`

Ebaneck commented on salt/metalk8s/addons/nginx-ingress-control-plane/pre-upgrade.sls:
```diff
@@ -0,0 +1,41 @@
+{%- set dest_version = pillar.metalk8s.cluster_version %}
+
+{#- In MetalK8s-2.4.2 we added selector in nginx ingress control plane
+    controller DaemonSet but `selector` are immutable
+    field so, in this case, we cannot replace the object we need to first
+    remove the current one and then deploy the desired one. #}
```
@TeddyAndrieux don't we have these selectors set for both control-plane and workload-plane ingress?

If so, we need to remove both objects.. no??

Is there an assumption here that selectors of `workload-plane` ingress are in good shape already before upgrade or downgrade??

TeddyAndrieux replied:
nginx ingress of `workload-plane` are done in `salt/metalk8s/addons/nginx-ingress/pre-upgrade.sls` here we are in `control-plane` one so we only take care of nginx ingress of control plane

Ebaneck replied:
yes, true!

Ebaneck reviewed (changes requested):
One question and a minor comment to resolve.

Should be good to merge soon

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort





The following reviewers are expecting changes from the author, or must review again:

* @ebaneck

bert-e commented:
# History mismatch



Merge commit #a6c8ca2f518e367e6eec4e2a9cb9d5e63a43d2d3 on the integration branch
`w/2.5/bugfix/add-pre-upgrade-downgrade-orchestrate` is merging a branch which is neither the current
branch `bugfix/add-pre-upgrade-downgrade-orchestrate` nor the development branch
`development/2.5`.

It is likely due to a rebase of the branch `bugfix/add-pre-upgrade-downgrade-orchestrate` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

TeddyAndrieux commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/2.5/bugfix/add-pre-upgrade-downgrade-orchestrate` with contents from `bugfix/add-pre-upgrade-downgrade-orchestrate`
and `development/2.5`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/2.5/bugfix/add-pre-upgrade-downgrade-orchestrate origin/development/2.5
 $ git merge origin/bugfix/add-pre-upgrade-downgrade-orchestrate
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.5/bugfix/add-pre-upgrade-downgrade-orchestrate
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort





The following reviewers are expecting changes from the author, or must review again:

* @ebaneck

TeddyAndrieux commented:
/approve

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort





The following reviewers are expecting changes from the author, or must review again:

* @ebaneck







*The following options are set:* **approve**

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.4`

* :heavy_check_mark: `development/2.5`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.4`

* :heavy_check_mark: `development/2.5`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`



Please check the status of the associated issue None.

Goodbye teddyandrieux.
